cfme/common/provider.py

```diff
@@ -207,21 +206,130 @@ def create(self, cancel=False, validate_credentials=True, check_existing=False,
+        if getattr(endpoint_default, "security_protocol", None):
+            security_protocol = endpoint_default.security_protocol.lower()
+            if security_protocol == "basic (ssl)":
+                security_protocol = "ssl"
+            default_connection["endpoint"]["security_protocol"] = security_protocol
+            con_config_to_include.append(default_connection)
+
+        # candu endpoint
+        if "candu" in self.endpoints:
+            endpoint_candu = self.endpoints["candu"]
+            if isinstance(prov_data["credentials"], dict):
+                prov_data["credentials"] = [prov_data["credentials"]]
+            prov_data["credentials"].append({
+                "userid": endpoint_candu.credentials.principal,
+                "password": endpoint_candu.credentials.secret,
+                "auth_type": "metrics",
+            })
+            candu_connection = {
+                "endpoint": {
+                    "hostname": endpoint_candu.hostname,
+                    "path": endpoint_candu.database,
+                    "role": "metrics",
+                },
+            }
+            if getattr(endpoint_candu, "api_port", None):
+                candu_connection["endpoint"]["port"] = endpoint_candu.api_port
+            if hasattr(endpoint_candu, "verify_tls") and not endpoint_candu.verify_tls:
+                candu_connection["endpoint"]["verify_ssl"] = 0
+            con_config_to_include.append(candu_connection)
+
+        prov_data["connection_configurations"] = []
+        appended = []
+        for config in con_config_to_include:
+            role = config["endpoint"]["role"]
+            if role not in appended:
+                prov_data["connection_configurations"].append(config)
+                appended.append(role)
 
-        logger.info('Setting up provider: %s via rest', self.key)
         try:
-            self.appliance.rest_api.collections.providers.action.create(
-                hostname=self.hostname,
-                ipaddress=self.ip_address,
-                name=self.name,
-                type="ManageIQ::Providers::{}".format(self.db_types[0]),
-                credentials={'userid': self.endpoints['default'].credentials.principal,
-                             'password': self.endpoints['default'].credentials.secret})
-
-            return self.appliance.rest_api.response.status_code == 200
-        except APIException:
-            return None
+            self.appliance.rest_api.collections.providers.action.create(**prov_data)
+        except APIException as err:
+            raise AssertionError("Provider wasn't added: {}".format(err))
+
+        response = self.appliance.rest_api.response
+        if not response:
+            raise AssertionError("Provider wasn't added, status code {}".format(
+                response.status_code))
+
+        if validate_inventory:
+            self.validate()
+
+        self.appliance.rest_api.response = response
+        return True
```

Comment:
**REQUIRED**
This method doesn't look good. It's too big. I would like to see it divided into several parts.